### PR TITLE
PR with multiple patches for st/nine

### PR DESCRIPTION
**Pull request by siro20**
Fixed version of "st/nine: Fix Reset when Resources weren't released properly"
Fixed version of "st/nine: Fix Swapchain dtor"

### CODE REVIEW

File: src/gallium/state_trackers/nine/swapchain9.c [hunk 1]
```diff
@@ -132,6 +131,12 @@ NineSwapChain9_Resize( struct NineSwapChain9 *This,
                 (pParams->SwapEffect == D3DSWAPEFFECT_COPY) ||
                 (pParams->SwapEffect == D3DSWAPEFFECT_DISCARD), D3DERR_INVALIDCALL);
 
+    if (pDevice->ex && pParams->BackBufferCount >= SWAPCHAIN_BUFFERS_MAX) {
```
Comment (axeldavy) [inline src/gallium/state_trackers/nine/swapchain9.c]: Why not just use these:
#define D3DPRESENT_BACK_BUFFERS_MAX    3
#define D3DPRESENT_BACK_BUFFERS_MAX_EX 30

If not ex return error when using more than D3DPRESENT_BACK_BUFFERS_MAX, and return error if using ex and more than D3DPRESENT_BACK_BUFFERS_MAX_EX buffers.
Use D3DPRESENT_BACK_BUFFERS_MAX_EX+1 slots for the table size.

**Comment by axeldavy** [comment]
It does indeed look like that we do support D3DPMISCCAPS_PERSTAGECONSTANT  (checking doc and code).

I don't like the first patch, but having hard time being inspired enough to find better.

File: src/gallium/state_trackers/nine/swapchain9.c (hunk 1 above)
Comment (siro20) [inline src/gallium/state_trackers/nine/swapchain9.c]: Why 30 back buffers? Where does this value come from?

Comment (axeldavy) [inline src/gallium/state_trackers/nine/swapchain9.c]: https://msdn.microsoft.com/en-us/library/windows/desktop/bb509549(v=vs.85).aspx

File: src/gallium/state_trackers/nine/device9.c [hunk 2]
```diff
@@ -120,6 +121,9 @@ NineDevice9_SetDefaultState( struct NineDevice9 *This, boolean is_reset )
     if (This->state.rs[D3DRS_ZENABLE])
         NineDevice9_SetDepthStencilSurface(
             This, (IDirect3DSurface9 *)This->swapchains[0]->zsbuf);
+
+    for (i = 1; i < This->caps.NumSimultaneousRTs; i++)
+        NineDevice9_SetRenderTarget(This, i, NULL);
```
Comment (axeldavy) [inline src/gallium/state_trackers/nine/device9.c]: render targets are already all released in nine_state_clear

**Comment by axeldavy** [comment]
wine tests say "A reference held to an implicit surface causes failures as well", I'm not sure with your first patch it will work for that part of the test.

**Comment by axeldavy** [comment]
gah I merged too fast, this doesn't work at all

**Comment by siro20** [comment]
What's wrong with it?

**Comment by axeldavy** [comment]
works only if an separate presentation buffer is used. I'm fixing it atm.

**Comment by axeldavy** [comment]
I fixed the bug, but I'm not sure how to introduce the fixes of this PR. Can you do a new patch on top of my fixes ?

**Pull request merged**